app/utility/base_knowledge_svc.py

```diff
@@ -0,0 +1,318 @@
+import copy
+import os
+import shutil
+import pickle
+import tarfile
+import uuid
+
+from datetime import datetime
+
+import app.service.data_svc
+from app.utility.base_service import BaseService
+
+DATA_BACKUP_DIR = app.service.data_svc.DATA_BACKUP_DIR
+FACT_STORE_PATH = "data/fact_store"
+
+
+class BaseKnowledgeService(BaseService):
+
+    def __init__(self):
+        self.log = self.add_service('knowledge_svc', self)
+        self.schema = dict(facts=[], relationships=[], rules=[], constraints=dict())
+        self.fact_ram = copy.deepcopy(self.schema)
+
+    # -- Fact API --
+    def _add_fact(self, fact, constraints=None):
+        """
+        Add a fact to the internal store
+        :param fact: Fact to add
+        :param constraints: any potential constraints
+        """
+        """facts can now be highly controlled, with visibility at the
+        operation level, agent(s) level, or custom groupings"""
+        if not any(x == fact for x in self.fact_ram['facts']):
+            fact._knowledge_id = uuid.uuid4()
+            self.fact_ram['facts'].append(fact)
+            if constraints:
```

Comment:
Are constraints conceptually defined yet? (I dont see them in design docs)

Reply:
No, that's just future proofing against the idea what we can add arbitrary tags/constraints to facts to specify when they can be used. They're not currently used for anything, but since we're planning what will hopefully be a stable API, it made sense to include the capability.